firedrake/slate/slac/utils.py

```diff
@@ -136,6 +152,224 @@ def visit_Symbol(self, o, *args, **kwargs):
+                        tensor_found = True
+            ret = Indexed(var, index)
+            ret = ComponentTensor(ret, index)
+        return ret
+
+    @slate_to_gem.register(firedrake.slate.slate.Add)
+    def slate_to_gem_add(self, tensor, node_dict):
+        A, B = tensor.operands  # slate tensors
+        _A, _B = node_dict[A], node_dict[B]  # gem representations
+        self.builder.create_index(A.shape, str(A)+"newadd"+str(B))
+        new_indices = self.builder.gem_indices[str(A)+"newadd"+str(B)]
+        _A = self.index_tensor(_A, new_indices)
+        _B = self.index_tensor(_B, new_indices)
+        return ComponentTensor(Sum(_A, _B), new_indices)
+
+    @slate_to_gem.register(firedrake.slate.slate.Negative)
+    def slate_to_gem_negative(self, tensor, node_dict):
+        A, = tensor.operands
+        self.builder.create_index(A.shape, str(A)+"newneg")
+        new_indices = self.builder.gem_indices[str(A)+"newneg"]
+        var_A = self.index_tensor(node_dict[A], new_indices)
+        return ComponentTensor(Product(Literal(-1), var_A), new_indices)
+
+    @slate_to_gem.register(firedrake.slate.slate.Transpose)
+    def slate_to_gem_transpose(self, tensor, node_dict):
+        A, = tensor.operands
+        _A = node_dict[A]
+        self.builder.create_index(A.shape, str(A)+"newtrans")
+        new_indices = self.builder.gem_indices[str(A)+"newtrans"]
+        var_A = self.index_tensor(_A, new_indices)
+        ret = ComponentTensor(var_A, new_indices[::-1])
+        return ret
+
+    @slate_to_gem.register(firedrake.slate.slate.Mul)
+    def slate_to_gem_mul(self, tensor, node_dict):
+        A, B = tensor.operands
+        var_A, var_B = node_dict[A], node_dict[B]  # gem representations
+
+        # New indices are necessary in case as Tensor gets multiplied with itself.
+        self.builder.create_index(A.shape, str(A)+"newmulA"+str(tensor))
+        new_indices_A = self.builder.gem_indices[str(A)+"newmulA"+str(tensor)]
+        self.builder.create_index(B.shape, str(B)+"newmulB"+str(tensor))
+        new_indices_B = self.builder.gem_indices[str(B)+"newmulB"+str(tensor)]
+
+        if len(A.shape) == len(B.shape) and A.shape[1] == B.shape[0]:
+            var_A = self.index_tensor(var_A, new_indices_A)
+            var_B = self.index_tensor(var_B, (new_indices_A[1], new_indices_B[1]))
+
+            prod = Product(var_A, var_B)
+            sum = IndexSum(prod, (new_indices_A[1],))
+            return ComponentTensor(sum, (new_indices_A[0], new_indices_B[1]))
+
+        elif len(A.shape) > len(B.shape) and A.shape[1] == B.shape[0]:
+            var_A = self.index_tensor(var_A, new_indices_A)
+            var_B = self.index_tensor(var_B, (new_indices_A[1],))
+
+            prod = Product(var_A, var_B)
+            sum = IndexSum(prod, (new_indices_A[1],))
+            return ComponentTensor(sum, (new_indices_A[0],))
+        else:
+            return[]
+
+    @slate_to_gem.register(firedrake.slate.slate.Block)
```

Comment:
```
firedrake.slate.slate.
firedrake.slate.slate.
firedrake.slate.slate.
firedrake.slate.slate.
```
The repeated use of fully qualified names is ugly.  Just use an abbreviation like in another file, e.g.
    import firedrake.slate.slate as sl

Reply:
Adressed in 84bdc32e0bf89c56175156951ebd3a939c5f89d1.